### PR TITLE
[Take-2] Infrastructure for Jax-based operators, IsingJax

### PR DESCRIPTION
**Pull request by PhilipVinc**
This builds on top of the code written originally by @wdphy16 and makes fully jax-compliant Ising operators as well as lays the groundwork for similar changes around...

replaces #1335 closes #1335

### CODE REVIEW

**Comment by wdphy16** [comment]
Maybe a broader topic: Do we have any plan to replace `Ising` and `Heisenberg` (or even `LocalOperator`) by `PauliStrings`? That may be a more general way to make them jittable

**Comment by PhilipVinc** [comment]
> Maybe a broader topic: Do we have any plan to replace Ising and Heisenberg (or even LocalOperator) by PauliStrings? That may be a more general way to make them jittable

`LocalOperator` cannot be replaced with straightforward `PauliString` only works for dimension-2 spaces.

Replacing Ising and possibly Heisenberg would be possible, indeed.
I actually think it would be a good idea.

Maybe a step-0 would be to add to all/most operators a `to_pauli_string` method that converts them, so we can start to more easily experiment with the tradeoffs.

In parallel, after this PR I would like to cleanup and merge your old `PauliStringJax` PR.

Once all this is done, we can asses the performance of all different approaches and decide what to do.

**Comment by PhilipVinc** [comment]
cc @inailuig what is missing from this PR is to make sure that `get_conn_padded` for `IsingJax` works fine for arbitrary shape/ndim inputs, and to implement `DiscreteOperatorJax.to_sparse`, possibly returning a jax sparse matrix, on top of `get_conn_padded` . Could you please take care of it?

File: netket/operator/_discrete_jax_operator.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    their own class from this class
+    """
+
+    @property
+    def max_conn_size(self) -> int:
+        """The maximum number of non zero ⟨x|O|x'⟩ for every x."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_conn_padded(self, x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (inailuig) [reply netket/operator/_discrete_jax_operator.py:37]: also If we ever want to do csr etc it migth make it easier

**Comment by inailuig** [comment]
> cc @inailuig what is missing from this PR is to make sure that `get_conn_padded` for `IsingJax` works fine for arbitrary shape/ndim inputs, and to implement `DiscreteOperatorJax.to_sparse`, possibly returning a jax sparse matrix, on top of `get_conn_padded` . Could you please take care of it?

both done. should now work with arbitrarily many batch dims

**Comment by PhilipVinc** [comment]
thanks. If tests pass I will merge this I think. Can you give an eye around the PR to spot if all is good in the PR?

**Comment by PhilipVinc** [comment]
Ok, with the last commit we also pick up a new way to compute the local energies using chunking that does not precompute all the connected elements anymore.

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -23,7 +23,7 @@ jobs:
           - os: ubuntu-latest
             python-version: "3.11"
             mpi: ""
-            doctest: true
+            doctest: false
```
Comment (gcarleo) [inline .github/workflows/CI.yml:26]: no doctest anymore?

Comment (PhilipVinc) [reply .github/workflows/CI.yml:26]: check the line above, we were running it on both python 3.10 and 3.11. This way we run standard tests on python 3.11 (and we keep the doctests on python 3.10).

I think it was a typo from a past PR.

File: netket/operator/_discrete_jax_operator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (inailuig) [reply netket/operator/_discrete_jax_operator.py]: nevermind, lets do this later and only for the operators where it might be relevant.

**Comment by PhilipVinc** [comment]
@inailuig does it look good for you?

**Review by inailuig — APPROVED** [review]
lgtm

**Pull request merged**